File: Difficulty: Medium/Subarrays With At Most K Distinct Integers/subarrays-with-at-most-k-distinct-integers.cpp
class Solution {
  public:
    int countAtMostK(vector<int> &arr, int k) {
        int n = arr.size();
        unordered_map<int, int> map;
        int start = 0, count = 0;
        
        for(int end=0; end<n; end++) {
            map[arr[end]]++;
            
            while(map.size() > k) {
                map[arr[start]]--;
                
                if(map[arr[start]] == 0)
                    map.erase(arr[start]);
                    
                start++;
            }
            
            count += (end - start) + 1;
        }
        
        return count;
    }
};

File: Difficulty: Medium/Subarrays With At Most K Distinct Integers/subarrays-with-at-most-k-distinct-integers.java
class Solution {
    public int countAtMostK(int arr[], int k) {
        int count = 0, i = 0, j = 0;
        HashMap<Integer, Integer> hm = new HashMap<>();
        
        while(i < arr.length) {
            hm.put(arr[i], hm.getOrDefault(arr[i], 0) + 1);
            while(hm.size() > k) {
               hm.put(arr[j], hm.getOrDefault(arr[j], 0) - 1); 
               if(hm.get(arr[j]) == 0) hm.remove(arr[j]);
               j++;
            }
            count += (i-j+1);
            i++;
        }
        
        return count;
    }
}
